controller/deploy_controller.rb

```diff
@@ -33,21 +39,21 @@ def environment_choice
       end
 
     when "test"
-      if !@other_tag_generator.tag_exists?("dev", @tags_for_this_commit)
-        @user_comms.tell_user_no_tag("dev")
-      elsif @other_tag_generator.tag_exists?("test", @tags_for_this_commit)
-        @user_comms.tell_user_already_a_tag("test")
+      if !@other_tag_generator.tag_exists?(dev, @tags_for_this_commit)
+        @user_comms.tell_user_no_tag(dev)
+      elsif @other_tag_generator.tag_exists?(test_tag, @tags_for_this_commit)
+        @user_comms.tell_user_already_a_tag(test_tag)
       else
-        apply_tag(@other_tag_generator.next_tag("dev", "test", @tags_for_this_commit))
+        apply_tag(@other_tag_generator.next_tag(dev_tag, test_tag, @tags_for_this_commit))
       end
 
     when "stage"
-      if !@other_tag_generator.tag_exists?("test", @tags_for_this_commit)
-        @user_comms.tell_user_no_tag("test")
-      elsif @other_tag_generator.tag_exists?("stage", @tags_for_this_commit)
-        @user_comms.tell_user_already_a_tag("stage")
+      if !@other_tag_generator.tag_exists?(test_tag, @tags_for_this_commit)
+        @user_comms.tell_user_no_tag(test_tag)
+      elsif @other_tag_generator.tag_exists?(stage_tag, @tags_for_this_commit)
+        @user_comms.tell_user_already_a_tag(stage_tag)
```

Comment:
This looks like quite a bit of duplication here, which you would need to further multiply when additional (hardcoded) environments are added. With the generalisation I mentioned above this would probably also disappear.